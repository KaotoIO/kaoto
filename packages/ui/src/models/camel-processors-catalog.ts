import { CatalogKind } from './catalog-kind';

export interface ICamelProcessorDefinition {
  model: ICamelProcessorModel;
  properties: {
    [K in string]: ICamelProcessorProperties;
  };
}

export interface ICamelProcessorModel {
  kind: CatalogKind.Processor;
  name: string;
  title: string;
  description?: string;
  deprecated: boolean;
  label: string;
  javaType?: string;
  abstract?: boolean;
  input?: boolean;
  output?: boolean;
}

export interface ICamelProcessorProperties {
  index: number;
  kind: string;
  displayName: string;
  required: boolean;
  type: string;
  javaType: string;
  oneOf: string[];
  deprecated: boolean;
  autowired: boolean;
  secret: boolean;
  description: string;
}
