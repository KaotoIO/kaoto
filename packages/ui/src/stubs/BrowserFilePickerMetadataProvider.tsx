import { ChangeEvent, createRef, FunctionComponent, PropsWithChildren, useCallback, useRef } from 'react';
import {
  SCHEMA_FILE_ACCEPT_PATTERN,
  SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY,
  SCHEMA_FILE_NAME_PATTERN,
  SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY,
} from '../models/datamapper';
import { IMetadataApi, MetadataContext } from '../providers';
import { readFileAsString } from './read-file-as-string';

export const BrowserFilePickerMetadataProvider: FunctionComponent<PropsWithChildren> = (props) => {
  const fileInputRef = createRef<HTMLInputElement>();
  const fileSelectionRef = useRef<{
    resolve: (files: Record<string, string>) => void;
    reject: (error: unknown) => unknown;
  }>();
  const fileContentsRef = useRef<Record<string, string>>();

  const askUserForFileSelection = useCallback(
    (
      include: string,
      _exclude?: string,
      _options?: Record<string, unknown>,
    ): Promise<string[] | string | undefined> => {
      if (!fileInputRef.current) return Promise.resolve(undefined);

      if (include === SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY;
      } else if (include === SCHEMA_FILE_NAME_PATTERN) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN;
      }

      fileInputRef.current.click();
      return new Promise<Record<string, string>>((resolve, reject) => {
        fileSelectionRef.current = { resolve, reject };
      }).then((files) => {
        fileContentsRef.current = files;
        return Object.keys(files);
      });
    },
    [fileInputRef],
  );

  const onImport = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const schemaFiles = event.target.files;
    if (!schemaFiles) return;
    const fileContents: Record<string, string> = {};
    const fileContentPromises: Promise<string>[] = [];
    Array.from(schemaFiles).forEach((f) => {
      const promise = readFileAsString(f).then((content) => (fileContents[f.name] = content));
      fileContentPromises.push(promise);
    });
    await Promise.allSettled(fileContentPromises);

    fileSelectionRef.current?.resolve(fileContents);
    fileSelectionRef.current = undefined;
    event.target.value = '';
  }, []);

  const getResourceContent = useCallback((path: string) => {
    return Promise.resolve(fileContentsRef.current && fileContentsRef.current[path]);
  }, []);

  const metadataApi: IMetadataApi = {
    askUserForFileSelection: askUserForFileSelection,
    getResourceContent: getResourceContent,
    shouldSaveSchema: true,
    getMetadata: () => Promise.resolve(undefined),
    setMetadata: () => Promise.resolve(),
    deleteResource: () => Promise.resolve(true),
    saveResourceContent: () => Promise.resolve(),
    getSuggestions: () => Promise.resolve([]),
    onStepUpdated: () => Promise.resolve(),
  };

  return (
    <MetadataContext.Provider value={metadataApi}>
      {props.children}
      <input
        type="file"
        style={{ display: 'none' }}
        data-testid={`attach-schema-file-input`}
        onChange={onImport}
        ref={fileInputRef}
      />
    </MetadataContext.Provider>
  );
};
