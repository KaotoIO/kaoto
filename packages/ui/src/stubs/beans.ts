export const beansYaml = `
- beans:
  - name: myBean
    type: io.kaoto.MyBean
    properties:
      p1: p1v
      p2: p2v
  - name: myBean2
    type: io.kaoto.MyBean
    properties:
      p1:
        p1s1: p1s1v
`

export const beansJson = {
  "beans": [
    {
      "name": "myBean",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": "p1v",
        "p2": "p2v"
      }
    },
    {
      "name": "myBean2",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": {
          "p1s1": "p1s1v"
        }
      }
    }
  ]
}

