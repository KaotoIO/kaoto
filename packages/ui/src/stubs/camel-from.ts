import { FromDefinition } from '@kaoto-next/camel-catalog/types';
import { CamelRouteVisualEntity } from '../models/visualization/flows';

/**
 * This is a stub Camel From in YAML format.
 * It is used to test the Canvas component.
 */
export const camelFromYaml = `
- from:
    uri: timer:tutorial
    steps:
    - to:
        uri: direct:my-route
`;

/**
 * This is a stub Camel From in JSON format.
 * It is used to test the Canvas component.
 */
export const camelFromJson: { from: FromDefinition } = {
  from: {
    uri: 'timer:tutorial',
    steps: [
      {
        to: {
          uri: 'direct:my-route',
        },
      },
    ],
  },
};

export const camelFrom = new CamelRouteVisualEntity({ from: camelFromJson.from });
