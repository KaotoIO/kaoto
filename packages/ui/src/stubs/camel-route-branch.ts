import { parse } from 'yaml';

export const camelRouteBranch = parse(`
- route:
    id: route-2768
    from:
      id: from-4014
      uri: timer:template
      parameters:
        period: "1000"
      steps:
        - choice:
            id: choice-3431
            otherwise:
              id: otherwise-3653
              steps:
                - log:
                    id: log-6808
                    message: \${body}
            when:
              - id: when-4112
                steps:
                  - setHeader:
                      id: setHeader-6078
                      expression:
                        simple: {}
                expression:
                  simple:
                    expression: \${header.foo} == 1
        - to:
            id: to-3757
            uri: sql
            parameters: {}
`);
