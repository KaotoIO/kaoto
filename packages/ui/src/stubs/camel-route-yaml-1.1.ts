export const camelRouteYaml_1_1_original = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        host: localhost
      steps: []
`;

export const camelRouteYaml_1_1_updated = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        bindingMode: "off"
        host: localhost
      steps: []
`;
