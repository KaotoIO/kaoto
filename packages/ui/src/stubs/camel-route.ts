import { RouteDefinition } from '@kaoto/camel-catalog/types';
import { parse } from 'yaml';

/**
 * This is a stub Camel Route in YAML format.
 * It is used to test the Canvas component.
 */
export const camelRouteYaml = `
- route:
    id: route-8888
    from:
      uri: timer
      parameters:
        timerName: tutorial
      steps:
      - set-header:
          simple: "\${random(2)}"
          name: myChoice
      - choice:
          when:
          - simple: "\${header.myChoice} == 1"
            steps:
            - log:
                id: log-1
                message: We got a one.
          otherwise:
            steps:
            - to:
                uri: 'amqp:queue:'
            - to:
                uri: 'amqp:queue:'
            - log:
                id: log-2
                message: "We got a \${body}"
      - to:
          uri: direct:my-route
          parameters:
            bridgeErrorHandler: true
`;

/**
 * This is a stub Camel Route in JSON format.
 * It is used to test the Canvas component.
 */
export const camelRouteJson: { route: RouteDefinition } = parse(camelRouteYaml)[0];

export const camelRouteWithDisabledSteps: { route: RouteDefinition } = parse(`
route:
  id: route-8888
  from:
    uri: timer
    steps:
      - log:
          disabled: true
          message: \${body}
      - to:
          uri: direct
          disabled: true
`);
