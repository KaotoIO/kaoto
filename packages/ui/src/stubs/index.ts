export * from './camel-route';
export * from './kamelet-binding-route';
export * from './kamelet-route';
export * from './pipe';
