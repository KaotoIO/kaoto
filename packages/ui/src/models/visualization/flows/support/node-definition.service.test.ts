import { beerSourceKamelet } from '../../../../stubs/beer-source-kamelet';
import { logModel } from '../../../../stubs/log-model';
import { xjTemplateAction } from '../../../../stubs/xj-template-action.kamelet';
import { ICamelProcessorProperty } from '../../../camel-processors-catalog';
import { NodeDefinitionService } from './node-definition.service';
import * as componentCatalogMap from '@kaoto-next/camel-catalog/camel-catalog-aggregate-components.json';

describe('NodeDefinitionService', () => {
  describe('getSchemaFromCamelCommonProperties', () => {
    it('should return an empty schema if the properties are undefined', () => {
      const schema = NodeDefinitionService.getSchemaFromCamelCommonProperties(undefined);
      expect(schema).toEqual({});
    });

    it('should return an empty schema if the properties are empty', () => {
      const schema = NodeDefinitionService.getSchemaFromCamelCommonProperties({});
      expect(schema).toEqual({ properties: {}, required: [], type: 'object' });
    });

    it('should return a schema with the properties', () => {
      const schema = NodeDefinitionService.getSchemaFromCamelCommonProperties(logModel.properties);

      expect(schema).toMatchSnapshot();
    });

    /* eslint-disable  @typescript-eslint/no-explicit-any */
    it('should return a schema for google-drive', () => {
      const catalog = (componentCatalogMap as any)['google-drive'];
      const schema = NodeDefinitionService.getSchemaFromCamelCommonProperties(catalog.properties);
      expect(schema.properties.scopes.items).toBeDefined();
    });
  });

  it('should return the JSON type for enums', () => {
    const enumProperty: ICamelProcessorProperty = {
      index: 1,
      kind: 'attribute',
      displayName: 'Library',
      required: false,
      type: 'enum',
      javaType: 'org.apache.camel.model.dataformat.AvroLibrary',
      enum: ['ApacheAvro', 'Jackson'],
      deprecated: false,
      autowired: false,
      secret: false,
      defaultValue: 'ApacheAvro',
      description: 'Which Avro library to use.',
    };

    const jsonType = NodeDefinitionService.getSchemaFromCamelCommonProperties({ enumProperty });
    expect(jsonType.properties.enumProperty.type).toEqual(undefined);
  });

  it('should return the JSON type for a duration field', () => {
    const durationProperty: ICamelProcessorProperty = {
      index: 1,
      kind: 'attribute',
      displayName: 'Batch Timeout',
      required: false,
      type: 'duration',
      javaType: 'java.lang.String',
      deprecated: false,
      autowired: false,
      secret: false,
      defaultValue: '1000',
      description: 'Sets the timeout for collecting elements to be re-ordered. The default timeout is 1000 msec.',
    };

    const jsonType = NodeDefinitionService.getSchemaFromCamelCommonProperties({ durationProperty });
    expect(jsonType.properties.durationProperty.type).toEqual('string');
  });

  it('should return the JSON type for a duration field', () => {
    const objectProperty: ICamelProcessorProperty = {
      index: 0,
      kind: 'expression',
      displayName: 'Correlation Expression',
      required: true,
      type: 'object',
      javaType: 'org.apache.camel.model.ExpressionSubElementDefinition',
      oneOf: ['python', 'xpath'],
      deprecated: false,
      autowired: false,
      secret: false,
      description: 'The expression used to calculate the correlation key to use for aggregation.',
    };

    const jsonType = NodeDefinitionService.getSchemaFromCamelCommonProperties({ objectProperty });
    expect(jsonType.properties.objectProperty.type).toEqual('object');
  });

  describe('getSchemaFromKameletDefinition', () => {
    it('should return an empty schema if the definition is undefined', () => {
      const schema = NodeDefinitionService.getSchemaFromKameletDefinition(undefined);
      expect(schema).toEqual({ properties: {}, required: [], type: 'object' });
    });

    it('should return a schema with the properties', () => {
      const schema = NodeDefinitionService.getSchemaFromKameletDefinition(beerSourceKamelet);

      expect(schema).toMatchSnapshot();
    });

    it('should mark the required properties in the schema', () => {
      const schema = NodeDefinitionService.getSchemaFromKameletDefinition(xjTemplateAction);

      expect(schema).toMatchSnapshot();
    });
  });
});
