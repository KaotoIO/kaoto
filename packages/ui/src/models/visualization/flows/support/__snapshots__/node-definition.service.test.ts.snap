// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeDefinitionService getSchemaFromCamelCommonProperties should return a schema with the properties 1`] = `
{
  "properties": {
    "description": {
      "default": undefined,
      "deprecated": false,
      "description": "Sets the description of this node",
      "title": "Description",
      "type": "string",
    },
    "disabled": {
      "default": false,
      "deprecated": false,
      "description": "Whether to disable this EIP from the route during build time. Once an EIP has been disabled then it cannot be enabled later at runtime.",
      "title": "Disabled",
      "type": "boolean",
    },
    "id": {
      "default": undefined,
      "deprecated": false,
      "description": "Sets the id of this node",
      "title": "Id",
      "type": "string",
    },
    "logName": {
      "default": undefined,
      "deprecated": false,
      "description": "Sets the name of the logger",
      "title": "Log Name",
      "type": "string",
    },
    "logger": {
      "default": undefined,
      "deprecated": false,
      "description": "To refer to a custom logger instance to lookup from the registry.",
      "title": "Logger",
      "type": "object",
    },
    "loggingLevel": {
      "default": "INFO",
      "deprecated": false,
      "description": "Sets the logging level. The default value is INFO",
      "enum": [
        "TRACE",
        "DEBUG",
        "INFO",
        "WARN",
        "ERROR",
        "OFF",
      ],
      "title": "Logging Level",
      "type": undefined,
    },
    "marker": {
      "default": undefined,
      "deprecated": false,
      "description": "To use slf4j marker",
      "title": "Marker",
      "type": "string",
    },
    "message": {
      "default": undefined,
      "deprecated": false,
      "description": "Sets the log message (uses simple language)",
      "title": "Message",
      "type": "string",
    },
  },
  "required": [
    "message",
  ],
  "type": "object",
}
`;

exports[`NodeDefinitionService getSchemaFromKameletDefinition should mark the required properties in the schema 1`] = `
{
  "properties": {
    "direction": {
      "description": "The transform direction.",
      "title": "Direction",
      "type": "string",
    },
    "template": {
      "description": "The inline template to apply a transformation through template.",
      "title": "Template",
      "type": "string",
    },
  },
  "required": [
    "direction",
    "template",
  ],
  "type": "object",
}
`;

exports[`NodeDefinitionService getSchemaFromKameletDefinition should return a schema with the properties 1`] = `
{
  "properties": {
    "period": {
      "description": "The time interval between two events",
      "title": "Period",
      "type": "integer",
    },
  },
  "required": [],
  "type": "object",
}
`;
